File: src/main/java/net/katsstuff/danmakucore/entity/spellcard/EntitySpellcard.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.entity.spellcard;

import java.util.Optional;
import java.util.UUID;

import javax.annotation.Nullable;

import net.katsstuff.danmakucore.entity.spellcard.spellcardbar.SpellcardInfoServer;
import net.katsstuff.danmakucore.helper.DanmakuHelper;
import net.katsstuff.danmakucore.helper.NBTHelper;
import net.katsstuff.danmakucore.misc.LogicalSideOnly;
import net.katsstuff.danmakucore.registry.DanmakuRegistry;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTUtil;
import net.minecraft.network.datasync.DataParameter;
import net.minecraft.network.datasync.DataSerializers;
import net.minecraft.network.datasync.EntityDataManager;
import net.minecraft.util.ResourceLocation;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.Side;

public class EntitySpellcard extends Entity {

	private static final DataParameter<Integer> SPELLCARD_ID = EntityDataManager.createKey(EntitySpellcard.class, DataSerializers.VARINT);

	private static final String NBT_SPELLCARD_TYPE = "spellcardType";
	private static final String NBT_SPELLCARD_DATA = "spellcardData";
	private static final String NBT_USER = "user";

	private boolean sendNamePacket = false;
	private SpellcardInfoServer spellcardInfo;

	@LogicalSideOnly(Side.SERVER)
	private EntityLivingBase user;
	@LogicalSideOnly(Side.SERVER)
	private SpellcardEntity spellCard;

	public EntitySpellcard(World world) {
		super(world);
		preventEntitySpawning = true;
		setSize(0.4F, 0.6F);
	}

	public EntitySpellcard(EntityLivingBase user, @Nullable EntityLivingBase target, Spellcard type, boolean sendInfoPacket) {
		this(user.world);

		setPosition(user.posX, user.posY + user.getEyeHeight(), user.posZ);
		setRotation(user.rotationYaw, user.rotationPitch);
		this.user = user;
		spellCard = type.instantiate(this, target);
		setSpellcardId(DanmakuRegistry.getId(Spellcard.class, type));
		this.sendNamePacket = sendInfoPacket;

		if(sendInfoPacket) {
			spellcardInfo = new SpellcardInfoServer(spellCard.getName());
		}
	}

	@Override
	public void onUpdate() {
		if(!world.isRemote && (spellCard == null || ticksExisted >= spellCard.type.getEndTime() || user == null || user.isDead)) {
			setDead();
			return;
		}

		super.onUpdate();

		if(spellcardInfo != null) {
			spellcardInfo.tick();
		}

		if(!world.isRemote) {
			spellCard.onUpdate();
			if(user instanceof EntityPlayer && ticksExisted < spellCard.type.getRemoveTime()) {
				DanmakuHelper.danmakuRemove(user, 40.0F, DanmakuHelper.DanmakuRemoveMode.OTHER, true);
			}
		}
	}

	@Override
	public void addTrackingPlayer(EntityPlayerMP player) {
		if(sendNamePacket) {
			spellcardInfo.addPlayer(player);
		}
	}

	@Override
	public void removeTrackingPlayer(EntityPlayerMP player) {
		if(sendNamePacket) {
			spellcardInfo.removePlayer(player);
		}
	}

	@Override
	protected void entityInit() {
		dataManager.register(SPELLCARD_ID, 0);
	}

	private void setSpellcardId(int number) {
		dataManager.set(SPELLCARD_ID, number);
	}

	public int getSpellcardId() {
		return dataManager.get(SPELLCARD_ID);
	}

	public void updateName() {
		if(spellcardInfo != null) {
			spellcardInfo.setName(spellCard.getName());
		}
	}

	@Override
	protected void readEntityFromNBT(NBTTagCompound tag) {
		UUID userUuid = NBTUtil.getUUIDFromTag(tag.getCompoundTag(NBT_USER));

		user = world.getPlayerEntityByUUID(userUuid);
		if(user == null) {
			Optional<Entity> optUser = NBTHelper.getEntityByUUID(userUuid, world);
			if(optUser.isPresent()) {
				Entity foundUser = optUser.get();
				if(foundUser instanceof EntityLivingBase) {
					user = (EntityLivingBase)foundUser;
				}
			}
		}

		if(user != null) {
			Spellcard type = DanmakuRegistry.SPELLCARD.getValue(new ResourceLocation(tag.getString(NBT_SPELLCARD_TYPE)));

			if(type == null) {
				setDead();
			}
			else {
				setSpellcardId(DanmakuRegistry.getId(Spellcard.class, type));
				spellCard = type.instantiate(this, user);
				spellCard.deserializeNBT(tag.getCompoundTag(NBT_SPELLCARD_DATA));
			}
		}
		else {
			setDead();
		}
	}

	@Override
	protected void writeEntityToNBT(NBTTagCompound tag) {
		tag.setString(NBT_SPELLCARD_TYPE, spellCard.type.getFullNameString());
		tag.setTag(NBT_SPELLCARD_DATA, spellCard.serializeNBT());
		tag.setUniqueId(NBT_USER, user.getUniqueID());
	}

	@Override
	public boolean canBePushed() {
		return false;
	}

	@Override
	public boolean canBeCollidedWith() {
		return false;
	}

	@Override
	public float getBrightness() {
		return 0.5F;
	}

	@LogicalSideOnly(Side.SERVER)
	public EntityLivingBase getUser() {
		return user;
	}

	@LogicalSideOnly(Side.SERVER)
	public SpellcardEntity getSpellCard() {
		return spellCard;
	}

	public Spellcard getSpellcardType() {
		return DanmakuRegistry.getObjById(Spellcard.class, getSpellcardId());
	}
}

File: src/main/scala/net/katsstuff/danmakucore/entity/spellcard/EntitySpellcard.scala
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.entity.spellcard

import net.katsstuff.danmakucore.entity.EntityInfo
import net.katsstuff.danmakucore.entity.spellcard.spellcardbar.SpellcardInfoServer
import net.katsstuff.danmakucore.helper.{NBTHelper, RemoveMode}
import net.katsstuff.danmakucore.lib.LibEntityName
import net.katsstuff.danmakucore.misc.LogicalSideOnly
import net.katsstuff.danmakucore.registry.DanmakuRegistry
import net.katsstuff.danmakucore.scalastuff.DanCoreImplicits._
import net.katsstuff.danmakucore.scalastuff.DanmakuHelper
import net.minecraft.entity.player.{EntityPlayer, EntityPlayerMP}
import net.minecraft.entity.{Entity, EntityLivingBase}
import net.minecraft.nbt.{NBTTagCompound, NBTUtil}
import net.minecraft.network.datasync.{DataSerializers, EntityDataManager}
import net.minecraft.util.ResourceLocation
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side

object EntitySpellcard {
  implicit val info: EntityInfo[EntitySpellcard] = new EntityInfo[EntitySpellcard] {
    override def create(world: World): EntitySpellcard = new EntitySpellcard(world)

    override def name: String = LibEntityName.SPELLCARD
  }

  private val SpellcardId = EntityDataManager.createKey(classOf[EntitySpellcard], DataSerializers.VARINT)
}
class EntitySpellcard(
    world: World,
    @LogicalSideOnly(Side.SERVER) var user: EntityLivingBase,
    target: Option[EntityLivingBase],
    @LogicalSideOnly(Side.SERVER) spellcardTpe: Spellcard,
    sendNamePacket: Boolean
) extends Entity(world) {
  private val NbtSpellcardType = "spellcardType"
  private val NbtSpellcardData = "spellcardData"
  private val NbtUser          = "user"

  @LogicalSideOnly(Side.SERVER)
  private var _spellcardEntity = if (spellcardTpe != null) {
    spellcardId = DanmakuRegistry.getId(classOf[Spellcard], spellcardTpe)
    spellcardTpe.instantiate(this, target)
  } else null

  preventEntitySpawning = true
  setSize(0.4F, 0.6F)

  if (user != null) {
    setPosition(user.posX, user.posY + user.getEyeHeight, user.posZ)
    setRotation(user.rotationYaw, user.rotationPitch)
  }

  @LogicalSideOnly(Side.SERVER)
  private val spellcardInfo =
    if (sendNamePacket && _spellcardEntity != null) new SpellcardInfoServer(_spellcardEntity.name) else null

  def this(world: World) = this(world, null, None, null, true)

  def this(user: EntityLivingBase, target: Option[EntityLivingBase], spellcard: Spellcard, sendInfoPacket: Boolean) {
    this(user.world, user, target, spellcard, sendInfoPacket)
  }

  override def onUpdate(): Unit = {
    if (!world.isRemote && (_spellcardEntity == null || ticksExisted >= spellcardTpe.endTime || user == null)) {
      setDead()
    } else {
      super.onUpdate()
      if (spellcardInfo != null) spellcardInfo.tick()

      if (!world.isRemote) {
        if(user.isDead) {
          DanmakuHelper.removeDanmaku(user, 40F, RemoveMode.Enemy, dropBonus = true)
          setDead()
        }
        else {
          _spellcardEntity.onUpdate()

          if (user.isInstanceOf[EntityPlayer] && ticksExisted < spellcardTpe.removeTime) {
            DanmakuHelper.removeDanmaku(user, 40.0F, RemoveMode.Other, dropBonus = true)
          }
        }
      }
    }
  }

  override def addTrackingPlayer(player: EntityPlayerMP): Unit = if (sendNamePacket) spellcardInfo.addPlayer(player)

  override def removeTrackingPlayer(player: EntityPlayerMP): Unit =
    if (sendNamePacket) spellcardInfo.removePlayer(player)

  override protected def entityInit(): Unit = dataManager.register(EntitySpellcard.SpellcardId, Int.box(0))

  private def spellcardId_=(number: Int): Unit = dataManager.set(EntitySpellcard.SpellcardId, Int.box(number))

  private def spellcardId: Int = dataManager.get(EntitySpellcard.SpellcardId)

  @LogicalSideOnly(Side.SERVER)
  def updateName(): Unit = if (spellcardInfo != null) spellcardInfo.setName(_spellcardEntity.name)

  override protected def readEntityFromNBT(tag: NBTTagCompound): Unit = {
    val userUuid = NBTUtil.getUUIDFromTag(tag.getCompoundTag(NbtUser))
    user = Option(world.getPlayerEntityByUUID(userUuid))
      .orElse(NBTHelper.getEntityByUUID(userUuid, world).toOption.collect { case living: EntityLivingBase => living })
      .orNull

    if (user != null) {
      val tpe = DanmakuRegistry.Spellcard.getValue(new ResourceLocation(tag.getString(NbtSpellcardType)))
      if (tpe == null) setDead()
      else {
        spellcardId = DanmakuRegistry.getId(classOf[Spellcard], tpe)
        _spellcardEntity = tpe.instantiate(this, user)
        _spellcardEntity.deserializeNBT(tag.getCompoundTag(NbtSpellcardData))
      }
    } else setDead()
  }

  override protected def writeEntityToNBT(tag: NBTTagCompound): Unit = {
    tag.setString(NbtSpellcardType, spellcardTpe.fullNameString)
    tag.setTag(NbtSpellcardData, _spellcardEntity.serializeNBT)
    tag.setUniqueId(NbtUser, user.getUniqueID)
  }

  override def canBePushed = false

  override def canBeCollidedWith = false

  override def getBrightness = 0.5F

  @LogicalSideOnly(Side.SERVER)
  def getUser: EntityLivingBase = user

  @LogicalSideOnly(Side.SERVER)
  def spellCardEntity: SpellcardEntity = _spellcardEntity

  @LogicalSideOnly(Side.SERVER)
  def getSpellCardEntity: SpellcardEntity = spellCardEntity

  def getSpellcardType: Spellcard = DanmakuRegistry.getObjById(classOf[Spellcard], spellcardId)
}
